### ISSUE TITLE
? Semantic Versioning Policy

### ISSUE THREAD

**Issue by MichaelDeBoey** [typescript-eslint/typescript-eslint#3320]
Hi team! 👋

When [looking at creating a semantic versioning policy for `eslint-plugin-testing-library`](https://github.com/testing-library/eslint-plugin-testing-library/discussions/343), we stubled uppon [ESLint's Semantic Versioning Policy](https://github.com/eslint/eslint#semantic-versioning-policy).  
When looking at the @typescript-eslint repo, we couldn't find one though.

As described in https://github.com/testing-library/eslint-plugin-testing-library/discussions/343#discussioncomment-664418, it would be awesome if you guys could tell us which versioning policy you're using, so we can take that into account when deciding what way we want to go.

**Comment by bradzacher** [comment]
More or less yeah that's the same policy we follow.

With linting it gets a bit fuzzy because you can fix a bug and make a rule report substantially more - which many people would call a breaking.

In general the policy for the plugin is:
- recommend sets only change in a breaking
- rules are only deleted in a breaking
- options are only removed in a breaking
- default options are only changed in a breaking
- increasing the minimum supported node or ESLint version only changes in a breaking

In general the policy for the parser is:
- nodes are only removed in a breaking
- properties are only removed in a breaking

For both - patch is generally reserved for bug fixes.

Then really everything else is fair game for a minor.

**Comment by MichaelDeBoey** [comment]
@bradzacher So reporting more errors with the same rule (so adding more cases) is that considered a breaking change?
Or would you guys first introduce this with a new option for the rule?

**Comment by bradzacher** [comment]
@MichaelDeBoey - it depends but generally no - not a breaking change.
I tend to evaluate each change based on what I think the users want.

If it's a change I think all users would want - then I don't consider the additional reports a problem or a breaking change.
If it's a change I think that some users wouldn't want - then I ask the author to gate the change behind a default off option, and then we'd turn it on in a breaking change.

Two example hypotheticals:

Say that our `no-unsafe-return` rule *just* checked to see if you were returning `any`, and then a PR added support for returning `Array<any>`/`Set<any>`/etc when you had a return type declared as `Array<NotAny>`/`Set<NotAny>`/etc.
I would **not** consider that PR to be a breaking change as no sane user would ever be against the extra reports - they holistically improve the rule.

A real example of breaking is #3273. I know that there are users purposely do comparisons against `never`, so enhancing the rule to report against those would report for those people when they don't want it. So that change would be very breaking without an option to gate it.

----

semver is a really tricky thing to create a set of hard-and-fast rules for. You need to tread a fine line to balance what you need to do to "move fast" against what your users expect.

For some projects what is "breaking" is really straightforward to define.
For linting it's really fuzzy. In the strictest sense any new lint report in an existing rule is a "breaking" change because a user won't be able to do a minor version upgrade without doing some manual changes.
In reality you can (and will) also introduce additional lint reports fixing pretty much any bug in a rule (if you've ever fixed a bug along the lines of "rule does not report in case X").

So yeah - ultimately define your "breaking" changes around what your users expect.

**Comment by Belco90** [comment]
Hey @bradleyayers, thanks for your explanation! It's really interesting.

Regarding those 2 examples you mentioned: don't you think those are covered by ESLint SemVer Policy?

1. That would be a minor release since it matches with "A bug fix in a rule that results in ESLint reporting more linting errors" or even with "A new addition to an existing rule to support a newly-added language feature (within the last 12 months) that will result in ESLint reporting more linting errors by default"
2. That would be a major release since it matches with "A new option to an existing rule that results in ESLint reporting more linting errors by default"

This is basically what you explained, but it seems they found some rule of thumb for defining it. What do you think?

**Comment by bradzacher** [comment]
My first example of `no-unsafe-return` would be classed as an enhancement to a rule - it's adding additional cases to an existing lint rule that were not previously within scope of the lint rule. It isn't a bug fix because it wouldn't be just a missed/forgotten case - it's an entirely new check being built into the rule.

So under ESLint's policy it'd fall as a breaking change "A new option to an existing rule that results in ESLint reporting more linting errors by default."

But I disagree with that.
As I mentioned - IMO semver with linting is about what users expect.
"Would the majority of users expect that `no-unsafe-return` also checks for unsafe generic assignments?"

This is where the fuzziness comes in - the answer to that question is based on my understanding of what  users expect gleaned from a few years of maintainership - years of hard learning by making the wrong decisions 😅 .

---

My second example could definitely be released as a breaking change - for sure, but because it's a change not all users would agree with - it's better to put it behind a default-off flag and release it in a minor.
Not all users would expect the rule to work that way, so making it work that way without an option is why I would class it as breaking.

---

This is the whole reason that TypeScript doesn't use semver at all - each release is a major release for them. It *looks* like they use semver because npm... but the fact that the current version is `4.2.4` means that this is the 42nd major release and the 4th release of that version.

Why don't they use semver? Simply put because for the TS project any change could technically be classed as breaking. Even a bugfix could cause TS to report or not report slightly different diagnostics.

**Comment by MichaelDeBoey** [comment]
@bradzacher I really like your way of thinking tbh.

The main reason why I wanted to investigate this a bit more in depth was because @Belco90 was already adhering to ESLint's versioning policy (as we were already fairly confident about that we wanted to do that), but to me it was becoming to feel that we couldn't improve any rules, except if we created a new option for each improvement (or if we released a major version each time) and that became to feel wrong for me tbh.

I think your way of thinking sheds a new light and perspective to things and will give us something extra to think about now.
